### PR TITLE
Improve ESP BLE and turn on BLE workflow

### LINKED ISSUE #9291: ESP32-S3 ADC use causes crashes when WiFi in use
### CircuitPython version

```python
1) Adafruit CircuitPython 9.0.5 on 2024-05-22; Waveshare ESP32-S3-Zero with ESP32S3
2) Adafruit CircuitPython 9.1.0-beta.3 on 2024-05-22; Waveshare ESP32-S3-Zero with ESP32S3
3) Adafruit CircuitPython 9.0.4 on 2024-04-16; Adafruit Feather ESP32-S3 TFT with ESP32S3
```

### Code/REPL

```python
import gc
import time
import board
import neopixel
from rainbowio import colorwheel
from adafruit_thermistor import Thermistor

led = neopixel.NeoPixel(board.NEOPIXEL, 1, brightness=0.1)

# Setup thermistor for readings
therm7 = Thermistor(board.A0, 10000, 10000, 25, 3695, high_side=True)  # pin, resistor, nom_thermistor, nom_temp
therm8 = Thermistor(board.A1, 10000, 10000, 25, 3695, high_side=True)

def get_average_temp(pin):
    readings = []

    for _ in range(5):
        reading = pin.temperature
        readings.append(reading)
        time.sleep(0.02)  # 20ms delay between readings

    average_temp_c = sum(readings) / len(readings)  # Average the C reading
    average_temp_f = (average_temp_c * 1.9) + 32  # Convert the Ave C Reading to F

    return average_temp_c, average_temp_f


# main loop
count = 0
while True:
    count += 1
    print(count, f"{gc.mem_alloc()=}")

    average_temp_c, average_temp_f = get_average_temp(therm7)
    print(
        f"   Average therm7 Reading : {average_temp_c:.0f}\u00b0C {average_temp_f:.0f}\u00b0F"
    )

    average_temp_c, average_temp_f = get_average_temp(therm8)
    print(
        f"   Average therm8 Reading : {average_temp_c:.0f}\u00b0C {average_temp_f:.0f}\u00b0F\n\n"
    )

    for x in range(3):
        led.fill(colorwheel((time.monotonic() * 50) % 255))  # change Neopixel color
        time.sleep(1)
```

### Behavior

Various failures but usually crashes share in common: MU pops up “Could not find an attached drive”, Mac OS pops up “Disk Not Ejected Properly”, MU of course has closed the serial window so nothing to see. Printing gc.mem_allocat() with each loop in my code shows allocated memorial in the 4000-8000 range so no apparent run away memory issues.

Sometimes the board will disconnect, come back, code stays running but the Neopixel is steady white like it is in the REPL. Other times it crashed with 3x yellow blinking (Safe mode) and reports an internal watchdog timer expired.

I have an S2 board that is on 9.0.4 and has been running this code for many weeks and sending the data to an IO feed. No chronic crashed like the S3 boards.

### Description

What follows is the long list of notes I have been taking as I tried different things. But the above, in behavior, is the executive summary. Below is tedious reading. Sorry...

Testing notes:

Waveshare ESP32-S3 Zero running 9.0.5 and libraries updated via Circup is starting with the “code chooser” code discussed here https://forums.adafruit.com/viewtopic.php?t=210926 starting with the 6th post down.

Code I am running (“choosing”) is a dual thermistor reading in a roughly 3+ second long loop that reads two thermistors and then changes the color of the Neopixel 3 times once per second. 

Crashes share in common: MU pops up “Count not find an attached drive”, Mac OS pops up “Disk Not Ejected Properly”, MU of course has closed the serial window so nothing to see. Printing gc.mem_allocat() with each loop in my code shows allocated memorial in the 4000-8000 range so no apparent run away memory issues.

First time I had no display configured so I could not see an error and the Neopixel wasn’t indicating any activity. Added code to display serial window on external display. Reset board with reset button. I failed to note if the drive had reloaded itself after the crash and before reseting the board.

Second time same “crash”. Observed the Neopixel to be constant white indicating it was in the REPL, however the external display showed the code was still running and getting valid thermistor readings. Crashed happened around 290-300 loops. CIRCUITPU remounted its drive I believe but not certain. I let it run for a while longer then reset the board with the reset button.

Third time crashed at loop 272, this time stopping and Neopixel flashing yellow in three blink bursts (safe mode). Reopening MU serial window, failed due to ”Internal watchdog timer expired.” Noted for sure that the CIRCUITPY drive had remounted. Ejected drive and power cycled the board by unplugging USB cable.

Fourth time crashed at loop 233 (gc.mem_alloc at 5568). Same as third run with code stopped, three yellow flashes, and “Internal watchdog timer expired” in the reopened MU serial window.

Switching gears… Renamed the “code chooser” program from code.py and made my thermistor code code.py so it will load and run directly without the chooser reseting the MCU. Also power cycle reset the board.

Different type of crash this time. At loop 53 (mem = 5168). Drive did not unmount and the error in the REPL is

Traceback (most recent call last):
  File "code.py", line 66, in <module>
  File "code.py", line 46, in get_average_temp
  File "adafruit_thermistor.py", line 126, in temperature
  File "adafruit_thermistor.py", line 116, in resistance
ZeroDivisionError: division by zero

Odd. Normally I use 10k resistors with my 10k thermistor but this time I only had 1k resistors on hand. But I wouldn’t think that should matter. Source code for the library doesn’t indicate any restrictions on the resistor range. I believe this failure is just a result of random values when no thermistor is attached.

Ran again and it made it to run 72 but same divide by zero error. Switched to 10k resistors. Hard reset. Made it to run 38, with the previously described crash scenario (disk eject & reconnect, safe mode with an “Internal watchdog timer expired” error) is back. Done for the night!

Next day. Backed up entire Waveshare CIRCUITPY drive. Ran one more time as is. Crashed with the Neopixel showing steady white (REPL indicator) but code was still running. MU and Mac OS both reported drive ejected. Board did not remount and MU doesn’t see it.

Adafruit REV TFT S2 Feather. Copied over all the files that were on the Waveshare. Also verified 9.0.5 and ran Circus to verify all libraries were up to date (all were). Commented out all code that had anything to do with the external display. Thermistors on breadboard changed from D6 and D7 to A0 and A1. No failures after a few hours.

Switched back to Waveshare and ran as is. Eventually failed with the REPL white neopixel, ejected disk, but kept running. Drive did not remount. Did full reinstall of boot loader then 9.0.5. Copied over backed up files onto the MCU again. Hard power cycle reset. Restarted code. Crashed at cycle 288, 3 yellow blink safe mode and “Internal watchdog timer expired” and drive remounted.

Commented out all thermistor stuff and just ran the neopixel and gc memory allocation. Ran 13908 loops without issue (over 12 hours). Uncommented thermistor code and restarted the run (hard reset). Made it 457 loops (a little over 20 minutes) and crashed with the board disconnecting and the TFT fade to black and back in about 3 second pulses.

Restarted as is after getting home from work. Got to about 275, white NeoPixel, still running code, and disconnected. Moved it to a power supply connection only (not computer) and restarted. Looks like it crashed the same way with white Neopixel and code still displaying new lines.

Copied same drive contents to S3 TFT Feather running 9.0.4 and started it on the computer (no thermistors connected). S3 TFT Feather crashed, disconnected, reconnected and reports Safe Mode for Internal Watchdog timer expired. Restarted S3 TFT Feather. Dies same way.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by tannewt**
The BLE workflow will broadcast publicly when the web workflow isn't activated.

EDIT by @dhalbert: fixes #9291 as well, due to the ESP-IDF upgrade.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -143,7 +143,7 @@
 [submodule "ports/espressif/esp-idf"]
 	path = ports/espressif/esp-idf
 	url = https://github.com/adafruit/esp-idf.git
-	branch = circuitpython-v5.1.3
+	branch = circuitpython-v5.2.2
```
[the remark this answers is not in the document]
Comment (dhalbert) [reply .gitmodules:146]: ADC issue in #9291 seems to be fixed by this 5.2.2 update! See https://github.com/adafruit/circuitpython/issues/9291#issuecomment-2161938330

**Comment by dhalbert** [comment]
Does this fix #9304?

**Comment by bill88t** [comment]
Fixes #9291.

BLE works, perfectly actually, however reloading with the ble workflow connected leads to a crash.
Only tested on ESP32-S3 so far.

IMG_2024-06-12-17-01-43-899

Will test on the rest of the chips.

File: .gitmodules (hunk 1 above)
[the remark this answers is not in the document]
Comment (tannewt) [reply .gitmodules:146]: I think we were using 5.2.1. We just failed to update this branch.

File: ports/espressif/common-hal/_bleio/Adapter.c [hunk 2]
```diff
@@ -499,6 +507,8 @@ uint32_t _common_hal_bleio_adapter_start_advertising(bleio_adapter_obj_t *self,
     if (ble_gap_adv_active() && !self->user_advertising) {
         return BLE_HS_EBUSY;
     }
+    // Override anonymous because it isn't working with the ESP-IDF.
+    anonymous = false;
```
[the remark this answers is not in the document]
Comment (tannewt) [reply ports/espressif/common-hal/_bleio/Adapter.c:511]: I'd rather not. Instead I'll file a bug. It should be fixable.

**Comment by tannewt** [comment]
> Does this fix #9304?

I haven't tried it. I was hoping @tyeth would test for me. I did fix some issues with characteristic current_value.

**Comment by tannewt** [comment]
> Some minor comments. This is still marked "Draft"; did you want to change that?

I was planning on undrafting once the CI is green.

File: supervisor/shared/bluetooth/serial.c [hunk 3]
```diff
@@ -36,10 +36,10 @@ const uint8_t circuitpython_base_uuid[16] = {0x6e, 0x68, 0x74, 0x79, 0x50, 0x74,
 static mp_obj_list_t characteristic_list;
 static mp_obj_t characteristic_list_items[3];
 
-static uint32_t _outgoing1[BLE_GATTS_VAR_ATTR_LEN_MAX / 4];
-static uint32_t _outgoing2[BLE_GATTS_VAR_ATTR_LEN_MAX / 4];
-static ble_drv_evt_handler_entry_t rx_static_handler_entry;
-static ble_drv_evt_handler_entry_t tx_static_handler_entry;
+static uint32_t _outgoing1[BLEIO_PACKET_BUFFER_MAX_PACKET_SIZE / 4];
+static uint32_t _outgoing2[BLEIO_PACKET_BUFFER_MAX_PACKET_SIZE / 4];
```
[the remark this answers is not in the document]
Comment (tannewt) [reply supervisor/shared/bluetooth/serial.c:40]: Added a macro to check.

**Comment by tannewt** [comment]
> BLE works, perfectly actually, however reloading with the ble workflow connected leads to a crash.
> Only tested on ESP32-S3 so far.

I've reproduced this and am looking now. Thanks!

**Comment by tyeth** [comment]
Should get a look at it tomorrow most likely (retesting #9304), but only got an ESP32v2 not s3 this time (away on vacation / working holiday).

**Comment by bill88t** [comment]
I did a little more testing.

Please, please, please, **please**, do not enable this for boards without psram.
ESP32-S3 doesn't have the memory for this..

Current main branch: 195k usable
This PR: 125k usable

ble takes 70k.  35% of all memory...
It will break a lot of applications.

If ble has to stay, can we please at least have a CIRCUITPY_NOBLE flag or something like this so that all the memory doesn't get allocated?
Or at least can we check as to if something is paired before taking all that ram?

It's actually catastrophic for me. I'm sweating at the though of this being merged.

**Comment by tannewt** [comment]
I was hoping that the RAM wouldn't be impacted until ble was imported. I'll take a look at it but may want this merged in the meantime.

**Comment by bill88t** [comment]
It may be `displayio` with the status bar is triggering the import.

**Comment by tannewt** [comment]
Looks like we could call `esp_bt_mem_release()` when ble hasn't been used and we're out of memory. That releases all of the BLE code so bleio can't be used afterwards.

**Comment by tyeth** [comment]
> Does this fix #9304?

Sorry not yet. Hard fault to safemode (tested on lilygo T-Display S3 which I forgot I'd brought with me). I ran your latest CI assets from this PR. 
I can see the values for accelerometer characteristic changing when viewed with Nrf Connect app.

**Review by dhalbert — APPROVED** [review]
Merging as is now to make it easier to test and to clear other merges.

**Pull request merged**